From 6c22fd93477ff632e5f259a0aa53cecf31908bc4 Mon Sep 17 00:00:00 2001
From: Henry Chen <henry.chen@oss.cipunited.com>
Date: Tue, 18 Aug 2026 17:39:22 +0800
Subject: [PATCH 77/77] loader/mips64/linux: work around memory exhaustion when
 loading linux

The default heap size isn't always sufficient when our kernels become
slightly bigger. Unlike 8c22693055cee21b3ab4af582e921fb71babc422,
I can't avoid using the relocator anymore, and this patch allows us
to drop that commit.

Signed-off-by: Henry Chen <henry.chen@oss.cipunited.com>
---
 grub-core/loader/mips64/linux.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/grub-core/loader/mips64/linux.c b/grub-core/loader/mips64/linux.c
index 6b6cac273..2231f4d68 100644
--- a/grub-core/loader/mips64/linux.c
+++ b/grub-core/loader/mips64/linux.c
@@ -113,6 +113,21 @@ find_mmap_size (void)
   return page_align (mmap_size);
 }
 
+// The relocator solely depends on grub mm's existing mmap for
+// out-of-memory detection, instead of directly searching the EFI
+// mmap. Let's try to shove half of all available pages into the
+// mm before the relocator complains.
+static int
+grub_efi_mmap_walk_hook (grub_uint64_t addr __attribute__((unused)), grub_uint64_t sz,
+			 grub_memory_type_t type, void *data __attribute__((unused)))
+{
+  if (type != GRUB_MEMORY_AVAILABLE)
+    return 0;
+  // ignore status; best-effort only
+  grub_mm_add_region_fn (sz >> 1, GRUB_MM_ADD_REGION_CONSECUTIVE);
+  return 0;
+}
+
 static grub_err_t
 grub_linux_boot (void)
 {
@@ -331,6 +346,7 @@ grub_linux_load32 (grub_elf_t elf, const char *filename)
 
   {
     grub_relocator_chunk_t ch;
+    grub_efi_mmap_iterate(grub_efi_mmap_walk_hook, NULL, 1);
     err = grub_relocator_alloc_chunk_addr (relocator, &ch,
 					   grub_vtop ((void *) target_addr),
 					   linux_size);
@@ -368,6 +384,7 @@ grub_linux_load64 (grub_elf_t elf, const char *filename)
 
   {
     grub_relocator_chunk_t ch;
+    grub_efi_mmap_iterate(grub_efi_mmap_walk_hook, NULL, 1);
     err = grub_relocator_alloc_chunk_addr (relocator, &ch,
 					   grub_vtop ((void *) target_addr),
 					   linux_size);
-- 
2.55.0

